core/reorder/rcm.cpp

```diff
@@ -0,0 +1,89 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2020, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/reorder/rcm.hpp>
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+#include "core/matrix/csr_kernels.hpp"
+#include "core/reorder/rcm_kernels.hpp"
+
+
+namespace gko {
+namespace reorder {
+namespace rcm {
+
+
+GKO_REGISTER_OPERATION(get_permutation, rcm::get_permutation);
+GKO_REGISTER_OPERATION(get_degree_of_nodes, rcm::get_degree_of_nodes);
+
+
+}  // namespace rcm
+
+
+template <typename ValueType, typename IndexType>
+void Rcm<ValueType, IndexType>::generate() const
+{
+    const IndexType num_rows = adjacency_matrix_->get_size()[0];
+    const auto exec = this->get_executor();
+    const auto mtx = adjacency_matrix_.get();
+    // RCM is only valid for symmetric matrices. Need to add an expensive check
+    // for symmetricity here ?
+    exec->run(rcm::make_get_degree_of_nodes(num_rows, mtx->get_const_row_ptrs(),
```

Comment:
I would personally also prefer passing a pointer to the buffer, but I guess that can be set aside. How do I go about calling that here, i. e. how to qualify that call, what to include? Does it matter we have a `SparsityCsr` here, and more specifically those possible diagonal entries should not be counted?

Reply:
But in case you meant that - yes, I'll remove it from the declaration.

Reply:
Good point, I did not think of that... That's basically the whole "Csr vs SparsityCsr" discussion again. Do you remember where we were on that point?

Reply:
Since we require `to_adjacency_matrix`, we need to use `SparsityCsr`. That could also be implemented for 'Csr`, or maybe only for it, and then we could move the `Csr` into a `SparsityCsr` (is that implemented?). For now, I think we can't quite de-duplicate this code.

Reply:
We could change the interface of `calculate_nonzeros_per_row` to take only row_ptrs and size, but that would complicate things again.

Reply:
I'll go ahead and close this here, but feel free to reopen if you want me to apply this change to the interface.